reactor-core/src/main/java/reactor/core/publisher/Flux.java

```diff
@@ -4029,6 +4029,27 @@ public final Flux<T> concatWith(Publisher<? extends T> other) {
 		return concat(this, other);
 	}
 
+	/**
+	 * If <a href="https://github.com/micrometer-metrics/context-propagation">context-propagation library</a>
+	 * is on the classpath, this is a convenience shortcut to capture thread local values during the
+	 * subscription phase and put them in the {@link Context} that is visible upstream of this operator.
+	 * <p>
+	 * As a result this operator should generally be used as close as possible to the end of
+	 * the chain / subscription point.
+	 * <p>
+	 * If context-propagation is not available at runtime, this operator simply returns the current {@link Flux}
+	 * instance.
+	 *
+	 * @return a new {@link Flux} where context-propagation API has been used to capture entries and
+	 * inject them into the {@link Context}
+	 */
+	public final Flux<T> contextCapture() {
+		if (!ContextPropagation.isContextPropagationAvailable()) {
+			return this;
```

Comment:
I think this situation deserves a log at `WARN` level. Otherwise, perhaps two methods should be available:
- `contextCaptureIfAvailable()` which fails silently,
- `contextCapture()` which logs `WARN` or even throws.

With the current implementation I'd expect someone to spend unnecessary amounts of time trying to debug this situation.

Reply:
Here is a summary of an internal discussion we had on this today.

It is a valid concern that `contextCapture()` may fail silently in cases where it is expected to work. However, at the an application cannot avoid direct use of the `context-propagation` if it wants to restore any `ThreadLocal` values. That means that in practice it will get a compilation issue in case of a classpath issue.

`contextCaptureIfAvailable` is a logical idea, but there are additional reasons anticipated to overload, e.g. by `Predicate` key and possibly others too, which leads to a multiplication of overloaded methods.

For the options that remain:

1. Change `contextCapture` to complain, most likely through an ISE (or log warning but those are more tricky from within a runtime call) but that would push the burden onto 3rd party libraries to use the operator conditionally. It's feasible but far from ideal, from an API perspective for what is an expected use case.
1. Require Micrometer `context-propagation`. Context propagation is a central topic for pretty much every reactive application, and most need to mix reactive and imperative at some point. It would be fitting then to make a stronger recommendation for this to always be present.
1. Defer the decision. For as long as an application cannot avoid direct use of the context-propagation API, the decision can be deferred which would also allow for more experience and feedback. We would see if we have more overloaded methods, whether the use of `context-propagation` becomes fully transparent, etc

Reply:
deferring that part. hopefully we get some feedback during RC1